mauth_client/middlewares/asgi.py

```diff
@@ -0,0 +1,91 @@
+import json
+import logging
+
+from asgiref.typing import (
+    ASGI3Application,
+    ASGIReceiveCallable,
+    ASGISendCallable,
+    Scope,
+)
+
+from mauth_client.authenticator import LocalAuthenticator
+from mauth_client.config import Config
+from mauth_client.consts import (
+    ENV_APP_UUID,
+    ENV_AUTHENTIC,
+    ENV_PROTOCOL_VERSION,
+)
+from mauth_client.signable import RequestSignable
+from mauth_client.signed import Signed
+from mauth_client.utils import decode
+
+logger = logging.getLogger("mauth_asgi")
+
+
+class MAuthASGIMiddleware:
+    def __init__(self, app: ASGI3Application, exempt: set = set()) -> None:
+        self._validate_configs()
+        self.app = app
+        self.exempt = exempt
+
+    async def __call__(
+        self, scope: Scope, receive: ASGIReceiveCallable, send: ASGISendCallable
+    ) -> None:
+        path = scope["path"]
+
+        if path in self.exempt:
+            return await self.app(scope, receive, send)
+
+        query_string = scope["query_string"]
+        url = f"{path}?{decode(query_string)}" if query_string else path
+        headers = {decode(k): decode(v) for k, v in scope["headers"]}
+        body = await self._get_body(receive)
+
+        signable = RequestSignable(
+            method=scope["method"],
+            url=url,
+            body=body,
+        )
+        signed = Signed.from_headers(headers)
+        authenticator = LocalAuthenticator(signable, signed, logger)
+        is_authentic, status, message = authenticator.is_authentic()
+
+        if is_authentic:
+            # asgi spec calls for passing a copy of the scope rather than mutating it
+            # note: deepcopy will blow up with infi recursion due to objects in some values
+            scope_copy = scope.copy()
+            scope_copy[ENV_APP_UUID] = signed.app_uuid
+            scope_copy[ENV_AUTHENTIC] = True
+            scope_copy[ENV_PROTOCOL_VERSION] = signed.protocol_version()
+            await self.app(scope_copy, receive, send)
+        else:
+            await self._send_response(send, status, message)
+
+    def _validate_configs(self) -> None:
+        # Validate the client settings (APP_UUID, PRIVATE_KEY)
+        if not all([Config.APP_UUID, Config.PRIVATE_KEY]):
+            raise TypeError("MAuthASGIMiddleware requires APP_UUID and PRIVATE_KEY")
+        # Validate the mauth settings (MAUTH_BASE_URL, MAUTH_API_VERSION)
+        if not all([Config.MAUTH_URL, Config.MAUTH_API_VERSION]):
+            raise TypeError("MAuthASGIMiddleware requires MAUTH_URL and MAUTH_API_VERSION")
```

Comment:
Instead of requiring `MAUTH_API_VERSION` can we not default to `v2`?

Reply:
i'd be ok with that

Reply:
ah, actually if you look in `config.py` file the version is actually optional and defaulted to `v1`

Reply:
i'd be ok with changing the default to `v2` pending others' thoughts.

Reply:
I heard we are going to stick with `v1` for a little more until `v3` comes out...

Reply:
ok, leaving "as-is" for now

Reply:
Seems good to me. Just wanted to make sure there was a sane default.